### PR TITLE
Feature/final methods

### PR DESCRIPTION
**Pull request by domjngo**

### CODE REVIEW

File: PrepareFiles.py [hunk 1]
```diff
@@ -1,106 +1,164 @@
...
         count = 1
         for file in data:
-            this_pdf.append(file['name'])
-            size = file['size'] + size
-            if size > max_deliveryfile_size:
-                output_file_list.append(this_pdf)
+            if length == count:
+                this_file.append(file['name'])
+                output_file_list.append(this_file)
+            elif (file['size'] + data[count]['size']) > max_deliveryfile_size:
```
Comment (milh0use) [inline PrepareFiles.py]: What are we testing here? I assume the logic should be "If the JPG file we are about to add 
to the PDF means that the maximum PDF file size will be exceeded, add it, but then start a new PDF", however we're getting the size of the file we're about to add (`file['size']`) adding the size of the next JPG in the list (`data[count]['size']`) and checking that it doesn't exceed max_deliveryfile_size?
Is there an error here?

File: PrepareFiles.py [hunk 2]
```diff
@@ -1,106 +1,164 @@
...
+                this_file.append(file['name'])
+                output_file_list.append(this_file)
                 size = 0
-                this_pdf = []
-            elif length == count:
-                output_file_list.append(this_pdf)
+                this_file = []
+            else:
+                this_file.append(file['name'])
```
Comment (milh0use) [inline PrepareFiles.py:77]: Also this else case looks like it's doing nearly the same thing?

File: PrepareFiles.py [hunk 3]
```diff
@@ -1,106 +1,164 @@
...
+    def _create_pdf(self, max_deliveryfile_size, reference, iaid):
+        batch_list = self._create_file_list(max_deliveryfile_size, self.replica_data['files'])
         output_name_prefix = self._create_file_name_prefix(reference)
         font = ImageFont.truetype('./font/Arial.ttf', 16)
         images = []
         n = 1
+        parts = []
+        count_images = 0
+        start = self._post_progress(iaid, 1)
+        if start:
```
Comment (milh0use) [inline PrepareFiles.py:141]: This if/else block is missing from the _create_zip def, should that do the same thing? If not, is there any point saving the outcome to the start variable in that def?

File: PrepareFiles.py (hunk 1 above)
Comment (domjngo) [reply PrepareFiles.py]: No error. I had an instance where one file was 49mb and the next file was 130mb. The previous logic would loop through once, then would say, that's fine it's under 50mb, then loop through the second time, add the next file of 130mb, and then would say, we are now over 50mb so finish this batch. But that means that batch would be 179mb!

File: PrepareFiles.py (hunk 2 above)
Comment (domjngo) [reply PrepareFiles.py:77]: Yeah, I think you're right. We probably don't need the below if statement anymore...

**Comment by gtvj** [comment]
Hi @domingobishop - I've gone through the code and it all seems OK to me. I've therefore approved the PR and you can merge it once you've made the changes @milh0use mentions above.

**Review by gtvj — APPROVED** [review]

File: PrepareFiles.py (hunk 3 above)
Comment (domjngo) [reply PrepareFiles.py:141]: This is reporting - back to the API and to CloudWatch. If this fails, I didn't think it should stop the process. Only log it in CloudWatch. Does this sound ok?

Comment (milh0use) [reply PrepareFiles.py:141]: I think so - the only thing I think is counterintuitive is that you do something different in the _create_pdf case than the _create_zip case. They should probably both do the same thing.

Comment (domjngo) [reply PrepareFiles.py:141]: Oh, yes, I missed that. I'll add it to the zip function.

Comment (domjngo) [reply PrepareFiles.py:141]: I actually missed a bunch of code. I've now added all need for progress reporting. It could be done probably better...

**Review by milh0use — APPROVED** [review]

**Pull request merged**